### ISSUE TITLE
[ MFTF ] Mess in ActionGroups: invalid names, multiple nodes.

### ISSUE THREAD

**Issue by lbajsarowicz** [magento/magento2#26331]
### Summary (*)
Goal of this ticket is to explain **Why** I put so much effort to cleanup the Magento's Funcionatl Testing Framework:

1. MFTF Best Practices states that Action Group name should be consistent with the filename (We're moving towards [having it as a requirement](https://github.com/magento/magento2-functional-testing-framework/issues/520))
2. Practical side: When looking for ActionGroup - it's enough if you search for a file with its "what we expect" name (eg. PHPStorm: `CTRL + SHIFT + N` and type `AdminLoginActionGroup`)
3. Maintainability: Only one ActionGroup with the same name is created in the module (when trying to create another one with the same name - you'll hit existing file).
4. Extensibility: I want to introduce `ActionGroup` lazy loading. When you're running even single Test, MFTF traverses all the files of ActionGroups to parse them and build the dependency of necessary files. I want to change that and use the only files that we need.

### Examples (*)
Report from my tool - for example for Customer module.

### Multiple ActionGroups in single file
```
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontOpenCustomerAccountCreatePageActionGroup.xml --- 2
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomerLogoutActionGroup.xml --- 2
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontAddCustomerAddressActionGroup.xml --- 2
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/SignUpNewUserFromStorefrontActionGroup.xml --- 13
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/OpenEditCustomerFromAdminActionGroup.xml --- 4
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/DeleteCustomerActionGroup.xml --- 2
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminEditCustomerAddressesFromActionGroup.xml --- 4
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminCustomerSubscribeNewsletterActionGroup.xml --- 2
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminCreateCustomerWithWebsiteAndStoreViewActionGroup.xml --- 2
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertCustomerSubscribeNewsletterActionGroup.xml --- 2
```

### Inconsistent ActionGroup names
```
/app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminSelectAllCustomersActionGroup.xml
        - AdminSelectAllCustomers
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertCustomerDefaultBillingAddressActionGroup.xml
        - AdminAssertCustomerDefaultBillingAddress
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminEditCustomerAddressNoZipNoStateActionGroup.xml
        - AdminEditCustomerAddressNoZipNoState
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminFilterCustomerGridByEmailActionGroup.xml
        - AdminFilterCustomerGridByEmail
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AssertCustomerLoggedInActionGroup.xml
        - AssertCustomerWelcomeMessageActionGroup
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertCustomerDefaultShippingAddressActionGroup.xml
        - AdminAssertCustomerDefaultShippingAddress
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontAssertSuccessLoginToStorefrontActionGroup.xml
        - StorefrontAssertSuccessLoginToStorefront
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertCustomerNoDefaultBillingAddressActionGroup.xml
        - AdminAssertCustomerNoDefaultBillingAddress
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminSelectCustomerByEmailActionGroup.xml
        - AdminSelectCustomerByEmail
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomerActionGroup.xml
        - CustomerLogoutStorefrontByMenuItemsActionGroup
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AssertStorefrontPasswordAutocompleteOffActionGroup.xml
        - AssertStorefrontPasswordAutoCompleteOffActionGroup
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomerElementNotVisibleActionGroup.xml
        - StorefrontCustomerReorderButtonNotVisibleActionGroup
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertNumberOfRecordsInCustomersAddressGridActionGroup.xml
        - AdminAssertNumberOfRecordsInCustomersAddressGrid
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/LoginToStorefrontWithEmailAndPasswordActionGroup.xml
        - LoginToStorefrontWithEmailAndPassword
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminCustomerGridActionGroup.xml
        - AdminFilterCustomerByEmail
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertCustomerAccountInformationActionGroup.xml
        - AdminAssertCustomerAccountInformation
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/NavigateCustomerActionGroup.xml
        - NavigateToAllCustomerPage
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminFilterCustomerByNameActionGroup.xml
        - AdminFilterCustomerByName
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminResetFilterInCustomerGridActionGroup.xml
        - AdminResetFilterInCustomerGrid
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomerAddressBookNotContainsActionGroup.xml
        - StorefrontCustomerAddressBookNotContains
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminResetFilterInCustomerAddressGridActionGroup.xml
        - AdminResetFilterInCustomerAddressGrid
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertCustomerGroupOnCustomerFormActionGroup.xml
        - AdminAssertCustomerGroupOnCustomerForm
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminSaveCustomerAndAssertSuccessMessageActionGroup.xml
        - AdminSaveCustomerAndAssertSuccessMessage
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/OpenStorefrontStoredPaymentMethodsPageActionGroup.xml
        - OpenStorefrontCustomerStoredPaymentMethodsPageActionGroup
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontRegisterCustomerFromOrderSuccessPageActionGroup.xml
        - StorefrontRegisterCustomerFromOrderSuccessPage
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/SetGroupCustomerActionGroup.xml
        - SetCustomerGroupForSelectedCustomersViaGrid
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminUpdateCustomerGroupActionGroup.xml
        - AdminUpdateCustomerGroupByEmailActionGroup
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertCustomerSubscribeNewsletterActionGroup.xml
        - AdminAssertCustomerIsSubscribedToNewsletters
        - AdminAssertCustomerIsSubscribedToNewslettersAndSelectedStoreView
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminCustomerSubscribeNewsletterActionGroup.xml
        - AdminSubscribeCustomerToNewsletters
        - AdminSubscribeCustomerToNewslettersAndSelectStoreView
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminEditCustomerAddressSetDefaultShippingAndBillingActionGroup.xml
        - AdminEditCustomerAddressSetDefaultShippingAndBilling
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertCustomerGroupPresentInGridActionGroup.xml
        - AdminAssertCustomerGroupPresentInGrid
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminConfigCustomerActionGroup.xml
        - SetCustomerDataLifetimeActionGroup
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontAssertRegistrationPageFieldsActionGroup.xml
        - StorefrontAssertRegistrationPageFields
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomerAddressBookNumberOfAddressesActionGroup.xml
        - StorefrontCustomerAddressBookNumberOfAddresses
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminCustomerSaveAndContinueActionGroup.xml
        - AdminCustomerSaveAndContinue
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/NavigateCustomerGroupActionGroup.xml
        - NavigateToCustomerGroupPage
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminDeleteAddressInCustomersAddressGridActionGroup.xml
        - AdminDeleteAddressInCustomersAddressGrid
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminEditCustomerAddressesFromActionGroup.xml
        - AdminEditCustomerAddressesFrom
        - AdminEditCustomerAddressSetDefaultShippingAndBilling
        - AdminEditCustomerAddressNoZipNoState
        - SelectDropdownCustomerAddressAttributeValueActionGroup
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminEditCustomerInformationFromActionGroup.xml
        - AdminEditCustomerAccountInformationActionGroup
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertAddressInCustomersAddressGridActionGroup.xml
        - AdminAssertAddressInCustomersAddressGrid
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/VerifyGroupCustomerActionGroup.xml
        - VerifyCustomerGroupForCustomer
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomerGoToSidebarMenuActionGroup.xml
        - StorefrontCustomerGoToSidebarMenu
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomerAddressBookContainsActionGroup.xml
        - StorefrontCustomerAddressBookContains
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertCustomerNoDefaultShippingAddressActionGroup.xml
        - AdminAssertCustomerNoDefaultShippingAddress
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/SwitchAccountActionGroup.xml
        - SignOut
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertCustomerGroupOnProductFormActionGroup.xml
        - AdminAssertCustomerGroupOnProductForm
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertCustomerInCustomersGridActionGroup.xml
        - AdminAssertCustomerInCustomersGrid
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminAssertErrorMessageCustomerGroupAlreadyExistsActionGroup.xml
        - AdminAssertErrorMessageCustomerGroupAlreadyExists
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminCustomerShopingCartActionGroup.xml
        - AdminAddProductToShoppingCartActionGroup
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontAddCustomerAddressActionGroup.xml
        - StorefrontAddNewCustomerAddressActionGroup
        - StorefrontAddCustomerDefaultAddressActionGroup
    /app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminFilterCustomerAddressGridByPhoneNumberActionGroup.xml
        - AdminFilterCustomerAddressGridByPhoneNumber
```

### Proposed solution
Currently I work with all the Open Source and Commerce modules to address the issue, as well as to report other issues - like workarounds applied to the tests, that create overhead when it comes to time of execution.

CC: @okolesnyk @dmytro-ch @ihor-sviziev @rogyar @soumyau @nuzil

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @dmytro-ch 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-30327` were created

**Issue Available**: @dmytro-ch, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by lbajsarowicz** [comment]
When this task is done, another one is ready to be done: https://github.com/magento/magento2-functional-testing-framework/issues/520

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz. Thank you for your report. 
The issue has been fixed in magento/magento2#25828 by @lbajsarowicz in 2.4-develop branch
Related commit(s):
 - [14f95d96d8c785bc4148ba9c4a27d4e84187dcbd](https://github.com/magento/magento2/commit/14f95d96d8c785bc4148ba9c4a27d4e84187dcbd)
 - [63582cd727fa8181daeae64f4de070f56de5630b](https://github.com/magento/magento2/commit/63582cd727fa8181daeae64f4de070f56de5630b)
 - [64b820df6a97a2a0b9cbe688f9552cd3e1fee797](https://github.com/magento/magento2/commit/64b820df6a97a2a0b9cbe688f9552cd3e1fee797)
 - [0652d0e5778d3cb4f69ecc9de95ebf85c38871b5](https://github.com/magento/magento2/commit/0652d0e5778d3cb4f69ecc9de95ebf85c38871b5)
 - [7ac6b48db1e4bfc614589fb5cbb58c1f8d289447](https://github.com/magento/magento2/commit/7ac6b48db1e4bfc614589fb5cbb58c1f8d289447)
 - [b5500f7552972882d352fc205214125ac51a2968](https://github.com/magento/magento2/commit/b5500f7552972882d352fc205214125ac51a2968)
 - [632b3e5fa8e10cd10c516e8b828237334d9fb059](https://github.com/magento/magento2/commit/632b3e5fa8e10cd10c516e8b828237334d9fb059)
 - [bf9315bbf482760afecffd2b355825debddd1689](https://github.com/magento/magento2/commit/bf9315bbf482760afecffd2b355825debddd1689)
 - [f7c59ba61f3a85f4b0da6ccf670513822beb33cc](https://github.com/magento/magento2/commit/f7c59ba61f3a85f4b0da6ccf670513822beb33cc)
 - [eaeacf18ba4133735d41e8212d0740c7b5bca768](https://github.com/magento/magento2/commit/eaeacf18ba4133735d41e8212d0740c7b5bca768)
 - [df511beac1bf058b86a2ca32b4aee50d96ab01cb](https://github.com/magento/magento2/commit/df511beac1bf058b86a2ca32b4aee50d96ab01cb)
 - [b7e5bcf3100f009f21a66b7dfa9b15ab8ba71931](https://github.com/magento/magento2/commit/b7e5bcf3100f009f21a66b7dfa9b15ab8ba71931)
 - [b7165b941ac4f6759563138b04cc0d667b7b932a](https://github.com/magento/magento2/commit/b7165b941ac4f6759563138b04cc0d667b7b932a)
 - [171215de56c2e08b396b3ef707afba7d84602112](https://github.com/magento/magento2/commit/171215de56c2e08b396b3ef707afba7d84602112)
 - [c6c52d63f9db8c48ec6795d1278c1dbf333e80d4](https://github.com/magento/magento2/commit/c6c52d63f9db8c48ec6795d1278c1dbf333e80d4)

The fix will be available with the upcoming 2.4.0 release.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz. Thank you for your report. 
The issue has been fixed in magento/magento2#26329 by @lbajsarowicz in 2.4-develop branch
Related commit(s):
 - [aa1271bbb9b1a813b456dae46976ced5b479d71c](https://github.com/magento/magento2/commit/aa1271bbb9b1a813b456dae46976ced5b479d71c)
 - [d013eff07517776415bab85953807ad87fbcd8a9](https://github.com/magento/magento2/commit/d013eff07517776415bab85953807ad87fbcd8a9)
 - [5cd51d42a27192e0159e2d46c9d273f7bf62053b](https://github.com/magento/magento2/commit/5cd51d42a27192e0159e2d46c9d273f7bf62053b)
 - [d1a5d56354bdacb71205ac54e3842b83301e5075](https://github.com/magento/magento2/commit/d1a5d56354bdacb71205ac54e3842b83301e5075)
 - [e48c617295e3510e24e5111153d5799210673a74](https://github.com/magento/magento2/commit/e48c617295e3510e24e5111153d5799210673a74)
 - [79797a7e9e30fd06f127d1983057f179d3f73280](https://github.com/magento/magento2/commit/79797a7e9e30fd06f127d1983057f179d3f73280)
 - [7c4b59b696e965426b47ff7ef82787687d07a08e](https://github.com/magento/magento2/commit/7c4b59b696e965426b47ff7ef82787687d07a08e)

The fix will be available with the upcoming 2.4.0 release.
